front/src/applications/operationalStudies/components/MacroEditor/ngeToOsrd.ts

```diff
@@ -258,155 +265,318 @@ const createTrainSchedulePayload = async ({
+      })
+    ).unwrap();
+    return { ...pacedTrain, id: timetableItemId };
+  }
+  const trainSchedule = await dispatch(
+    osrdEditoastApi.endpoints.getTrainScheduleById.initiate({
+      id: formatTrainScheduleIdToEditoastTrainId(timetableItemId),
+    })
+  ).unwrap();
+  return { ...trainSchedule, id: timetableItemId };
+};
+
+const createPacedTrain = async (
+  timetableId: number,
+  pacedTrain: PacedTrainBase,
+  dispatch: AppDispatch
+): Promise<PacedTrainResultWithPacedTrainId> => {
+  const newTimetableItems = await dispatch(
+    osrdEditoastApi.endpoints.postTimetableByIdPacedTrains.initiate({
+      id: timetableId,
+      body: [pacedTrain],
+    })
+  ).unwrap();
+  if (!newTimetableItems || newTimetableItems.length === 0) {
+    throw new Error('Failed to create paced train');
+  }
   return {
-    train_name: trainrun.name,
-    labels: compact(trainrunLabels),
-    path: path.flat(),
-    start_time: startDate.toISOString(),
-    schedule,
+    ...newTimetableItems[0],
+    id: formatEditoastTrainIdToPacedTrainId(newTimetableItems[0].id),
   };
 };
 
-const handleUpdateTrainSchedule = async ({
+const updateTimetableItem = async (
+  timetableItemId: TimetableItemId,
+  timetableItem: TimetableItemWithTimetableId,
+  dispatch: AppDispatch
+): Promise<TimetableItemWithTimetableId> => {
+  if (isPacedTrain(timetableItemId)) {
+    const newPacedTrain = await dispatch(
+      osrdEditoastApi.endpoints.putPacedTrainById.initiate({
+        id: formatPacedTrainIdToEditoastTrainId(timetableItemId),
+        pacedTrainForm: {
+          ...(timetableItem as PacedTrainResultWithPacedTrainId),
+          // Reset margins because they contain references to path items
+          margins: undefined,
+        },
+      })
+    ).unwrap();
+    return { ...newPacedTrain, id: timetableItemId };
+  }
+  const newTrainSchedule = await dispatch(
+    osrdEditoastApi.endpoints.putTrainScheduleById.initiate({
+      id: formatTrainScheduleIdToEditoastTrainId(timetableItemId),
+      trainScheduleForm: {
+        ...(timetableItem as TrainScheduleResultWithTrainId),
+        // Reset margins because they contain references to path items
+        margins: undefined,
+      },
+    })
+  ).unwrap();
+  return { ...newTrainSchedule, id: timetableItemId };
+};
+
+const handleCreateTimetableItem = async (
+  timetableId: number,
+  netzgrafikDto: NetzgrafikDto,
+  trainrun: TrainrunDto,
+  infraId: number,
+  dispatch: AppDispatch,
+  addUpsertedTimetableItems: (timetableItems: TimetableItemWithTimetableId[]) => void,
+  state: MacroEditorState
+): Promise<void> => {
+  const timetableItemBase = await generateTimetableItemBase({
+    trainrunSections: getTrainrunSectionsByTrainrunId(netzgrafikDto, trainrun.id),
+    nodes: netzgrafikDto.nodes,
+    trainrun,
+    infraId,
+    dispatch,
+    labels: netzgrafikDto.labels,
+  });
+  const newPacedTrain = await createPacedTrain(
+    timetableId,
+    timetableItemBase as PacedTrainBase,
```

Comment:
We try to avoid casting. 
```suggestion
    timetableItemBase,
```

To remove this one, you can add a special type definition on `generateTimetableItemBase`:
```
type TimetableItemBaseProps = {
  trainrunSections: TrainrunSectionDto[];
  nodes: NodeDto[];
  trainrun: TrainrunDto;
  infraId: number;
  dispatch: AppDispatch;
  labels: LabelDto[];
};

async function generateTimetableItemBase(props: TimetableItemBaseProps): Promise<PacedTrainBase>;
async function generateTimetableItemBase(
  props: TimetableItemBaseProps & {
    timetableItem?: TimetableItemWithTimetableId;
    trainrunFrequencies?: TrainrunFrequency[];
  }
): Promise<TimetableItemBase>;
async function generateTimetableItemBase({
  trainrunSections,
  nodes,
  trainrun,
  infraId,
  dispatch,
  labels,
  timetableItem,
  trainrunFrequencies,
}: TimetableItemBaseProps & {
  timetableItem?: TimetableItemWithTimetableId;
  trainrunFrequencies?: TrainrunFrequency[];
}): Promise<TimetableItemBase>
```
This means that if you don't give a timetableItem, you will receive at the end a PacedTrainBase. In the other case, you will receive a TimetableItemBase

Reply:
Thanks a lot, I wouldn't have come up with this without your help. Done in "refactor update to reduce casting"